packages/matter.js/src/cluster/server/ClusterServer.ts

```diff
@@ -131,10 +133,20 @@ export function ClusterServer<
             }
 
             if (datasource.eventHandler) {
+                const promises = new Array<PromiseLike<void>>();
                 for (const eventName in events) {
-                    (events as any)[eventName].bindToEventHandler(datasource.eventHandler);
+                    const bindResult = (events as any)[eventName].bindToEventHandler(datasource.eventHandler);
+                    if (bindResult !== undefined && MaybePromise.is(bindResult)) {
+                        promises.push(bindResult);
+                    }
+                }
+                if (promises.length > 0) {
+                    return Promise.all(promises).then(() => Promise.resolve()) as S extends SyncStorage
```

Comment:
Whoo boy I know why you have to cast a Promise to void but that's nasty looking.  What if you create a utility type such as:

```js
type StorageOperationResult<S extends SyncStorage | AsyncStorage, T = void> =
    S extends SyncStorage ? T : Promise<T>
```